src/centralized_telescope/algorithm.rs

```diff
@@ -4,21 +4,26 @@ use super::proof::Proof;
 use super::round::Round;
 use super::setup::Setup;
 use super::types::Hash;
-use crate::utils::sample;
-use crate::utils::types::Element;
+use crate::utils::{sample, types::ToBytes};
 use blake2::{Blake2s256, Digest};
 
 /// Alba's proving algorithm, based on a depth-first search algorithm.
 /// Calls up to setup.max_retries times the prove_index function and returns an empty
 /// proof if no suitable candidate is found.
-pub fn prove(setup: &Setup, prover_set: &[Element]) -> Option<Proof> {
+pub fn prove<Element>(setup: &Setup, prover_set: &[Element]) -> Option<Proof<Element>>
```

Comment:
@rrtoledo @tolikzinovyev @curiecrypt : if I think about it, the most general setting occurs when `prove()` is parameterized by a generic type and this generic type under the constraint that this type implemetns the traits Eq + Hash. Can we do something like this 
```
fn prove<T: Eq + Hash>()
```

Reply:
If you don't want to write `<T: Eq + Hash>` in every function signature, you can first define 
```
use std::hash::Hash;
use std::collections::HashSet;

trait HashableEq: Eq + Hash {}

impl<T: Eq + Hash> HashableEq for T {}
```
This implements `HashableEq` for all types `T` that requires implementing both `Eq` and `Hash` and the second line ensures that any type that implements Eq and Hash automatically satisfies `HashableEq`, so you can now write 
```
fn prove<T: HashableEq>(...) {}
```

Reply:
@djetchev This looks interesting, we would also need the Copy trait however.

I had a look at the `Hash` trait in the past, and I wanted to stay clear from it as it does not seem it was meant for cryptographic purposes (but DoS for hashmaps).

If you look at [how it works](https://doc.rust-lang.org/std/hash/trait.Hash.html), you need to use a `Hasher` in conjunction to it. However the [default hasher](https://github.com/rust-lang/rust/blob/master/library/std/src/hash/random.rs#L110) is SipHasher13 which is not cryptographically secure but used only to prevent DOS in hashmaps.

I am afraid that by using this trait, we are encourage the use of this default hasher. Perhaps I am being overlycautious, and we can simply provide a different default hash.